---
title: "Mannitol transgene and salt stress - in plants."
author: "Jennifer Daly, Micaela Beckman"
date: "April 24, 2017"
output: html_document
---
INTRODUCTION
In this experiment, Arabidopsis plants were transformed with the mannose-6-phosphate reductase (M6PR) gene from celery. These plants were observed to be more salt tolerant. They exibited reduced salt injury, had more vegitative growth and more seed production in the presence of salt than the wild-type (COL) plants. 

In the presence of salt, many fewer gene differences were measured between M6PR and wild type (col) plants, this suggests that the M6PR gene was successful in pre-conditioning these plants to salt. In the untreated plants, many more genes were differently expressed. 

If in our analysis, these results are the same in that in the presence of salt less genes show differences, we can conclude that the original paper is accurate. 

To measure this, we will define three groups: 
(A) Pathways mainly affected by salt stress
(B) Pathways mainly activated by the M6PR transgene
(C) Pathways affected by both salt stress and the M6PR transgene

H0: Our results are synonymous with the paper, group A has a lower number of affected genes than any other group. 
Ha: Our results differ from the paper in some way. 

In the paper, they used Bioconductor's GCRMA to normalize the data, AffyLimGUI to create p-values and linear plots of the data, and Matman to tie the probe sets to annotations. 

If you were to attempt to re-create exactly how the data was analyzed in the paper, you would install Bioconductor and all packages as instructed in this link: http://bioconductor.org/install/

In order to install and use AffyLimGUI, we attempted to follow the instructions on these pages:
Basic install & run guide: http://bioinf.wehi.edu.au/affylmGUI/#install

A run-through of specific features using a data set: https://www.bioconductor.org/packages/release/bioc/vignettes/affylmGUI/inst/doc/estrogen/estrogen.html

We were able to install all dependencies for AffyLimGUI on Windows, but when attempting to create a plot we got an error "Could not find function 'indexProbes'." We could not find proper help on this issue, so we decided to not use the GUI in our analysis. We were still able to use GCRMA for normalization in our analysis, and we chose to use both GCRMA and RMA to compare. 

ANALYSIS:
Begin Analysis, data import & normalization
```{r, message=FALSE, warning=F}
#Required libraries
library(affy)
library(limma)
library(RSQLite)
library(gcrma)

#Import data from sources
pd = read.AnnotatedDataFrame("samples.txt",header=TRUE,sep=",",row.names=1) 
mydata = ReadAffy(filenames=pd$filename,phenoData=pd,verbose=TRUE)
sampleNames(mydata)=row.names(pData(pd))
#Call functions to organize data
eset_rma = rma(mydata)
eset_gc = gcrma(mydata)
expression_data = exprs(eset)
```

The difference between RMA and GCRMA is that GCRMA is based on a model using GC content first, while RMA uses a convolution model first for background correction. It is understood that this causes GCRMA to attain more accurate gene expression. 
Reference: https://www.stat.berkeley.edu/~terry/Classes/s246.2006/Week10/Week10L1.pdf 

*A Note: During our image check, we found that COL.U1 has multiple 'blobs'. We saw this on both computers and this may bias the results, but we are not sure in what way.* 

Create a design matrix: Set column labels
```{r}
TS = paste(pd$genotype,pd$treatment,sep=".")
TS = factor(TS, levels= c("M6PR.U","M6PR.T","COL.U","COL.T"))
```

Create a design matrix: Populate the columns by creating a list who's entries are 1/0 values based on the experiment names. 
```{r}
exp_design = model.matrix(~0 + TS)
colnames(exp_design) = levels(TS)
exp_design
```

Fit a linear model to the data:
*This is where we would have used AffyLimGUI, had we be able to*
```{r}
fit=lmFit(eset_rma,exp_design)
gcfit=lmFit(eset_gc,exp_design)
```

We chose to begin our analysis with a bar graph representation of the data because it was easier to compare with the original paper this way. In our bar graphs, the GCRMA graph is more similar than the RA graph in that the control untreated vs treated has the highest differences in gene expression, and the M6PR treated vs. untreated has the lowest. This is synonymous with the paper's conclusions that this gene pre-conditions the plant to salt, and so in the presense of salt, the plant exibits less differences in gene expression. 

Functions for creating a bar graph with RMA normalization:
```{r}
cont.matrix= makeContrasts(
    M6PR.TvU = M6PR.U - M6PR.T,
    COL.UvT = COL.U - COL.T,
    M6PRvCOL.T = M6PR.T - COL.T,
    M6PRvCOL.U = M6PR.U - COL.U,
    levels = exp_design)
fit2 = contrasts.fit(fit,cont.matrix)
fit2 = eBayes(fit2)
```

Functions for creating a bar graph with GCRMA normalization:
```{r}
cont.matrix2= makeContrasts(
    M6PR.TvU = M6PR.U - M6PR.T,
    COL.UvT = COL.U - COL.T,
    M6PRvCOL.T = M6PR.T - COL.T,
    M6PRvCOL.U = M6PR.U - COL.U,
    levels = exp_design)
gcfit2 = contrasts.fit(gcfit,cont.matrix2)
gcfit2 = eBayes(gcfit2)
```

Upon running GCRMA, we get a warning message "Zero sample variences detected, have been offset away from zero". We found that it does this so that when using statistical tests we won't have zero values in the denominator which would cause infinite result values. Reference: https://stat.ethz.ch/pipermail/bioconductor/2007-May/017025.html

Create a bar graph for the rma normalized data:
```{r}
# lfc =1 for 2-fold changes
result = decideTests(fit2, p.value=0.05, lfc=1)

colnames = c("M6PR.TvU",	"COL.UvT",	"M6PRvCOL.T",	"M6PRvCOL.U")
rownames = c("upregulated", "downregulated")
upregulated = c(sum(result[,1]>0), sum(result[,2]>0), sum(result[,3]>0), sum(result[,4]>0))
downregulated = c(sum(result[,1]<0), sum(result[,2]<0), sum(result[,3]<0), sum(result[,4]<0))

resultsMatrix = matrix(c(upregulated, downregulated), nrow=2, dimnames=list(rownames, colnames))
barplot(as.matrix(resultsMatrix), legend = rownames, beside=TRUE)
```

This bar graph result is very different from the paper. The two treated groups contain the highest amount of varience, suggesting that the M6PR salt-resistance gene had no effect. 
Figure for reference: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3192998/figure/fig5/

Create a bar graph for the GCRMA normalized data:
```{r}
# lfc =1 for 2-fold changes
gc_result = decideTests(gcfit2, p.value=0.05, lfc=1)

gc_colnames = c("M6PR.TvU",	"COL.UvT",	"M6PRvCOL.T",	"M6PRvCOL.U")
gc_rownames = c("upregulated", "downregulated")
gc_upregulated = c(sum(gc_result[,1]>0), sum(gc_result[,2]>0), sum(gc_result[,3]>0), sum(gc_result[,4]>0))
gc_downregulated = c(sum(gc_result[,1]<0), sum(gc_result[,2]<0), sum(gc_result[,3]<0), sum(gc_result[,4]<0))

gc_resultsMatrix = matrix(c(gc_upregulated, gc_downregulated), nrow=2, dimnames=list(gc_rownames, gc_colnames))
barplot(as.matrix(gc_resultsMatrix), legend = rownames, beside=TRUE)
```

Bar graph key:
(A) Pathways mainly affected by salt stress: COL.UvT
(B) pathways mainly activated by the M6PR transgene: M6PRvCOL.U
(C) pathways affected by both salt stress and the M6PR transgene: M6PRvCOL.T
(D) Pathways affected by the presence of the M6PR gene: M6PR.TvU

This bar graph result was a close match to the results show in the paper. A major difference we see between our bar graph and theirs is the amount of up regulation. It is possible that when the data was offset from 0 it affected how we measure up or down regulated genes. If the values were offset from 0 and we are determining up or down regulation by variation from 0, then the actual amount of up/down regulated genes is different from what we see here. 

We chose to only include the GCRMA data in the venn diagram analysis because this method of normalization is more accurate, and it is a closer fit to the paper's data. 

Functions for creating a venn diagram:
```{r}
cont.matrix3= makeContrasts(
    M6PR.TvU = M6PR.U - M6PR.T,
    COL.UvT = COL.U - COL.T, 
    Diff = (M6PR.U - M6PR.T) - (COL.U - COL.T),
    levels = exp_design)
gcfit3 = contrasts.fit(gcfit,cont.matrix3)
gcfit3 = eBayes(gcfit3)
```

Create a Venn Diagram for the GCRMA normalized data:
```{r}
# lfc =1 for 2-fold changes
gc_result2 = decideTests(fit3, p.value=0.05, lfc=1)
vennDiagram(gc_result2)
vennCounts(gc_result2)
```

Venn Diagram key:
(A) Pathways mainly affected by salt stress; COL.UvT
(B) Pathways mainly activated by the M6PR transgene; Differences between M6PR.TvU and COL.UvT (24, 20)
(C) Pathways affected by both salt stress and the M6PR transgene; Diff

In our analysis, there are thousands more expressed genes than the paper included in their analysis. We assume that this is because we could not use AffyLimGUI. We used the same data-choosing methods used in the paper, we only included probe sets with a p-value higher than 0.05 and greater than 2 log-fold change. 

Retrieve annotations:
```{r}
f="ftp://ftp.arabidopsis.org/Microarrays/Affymetrix/affy_ATH1_array_elements-2010-12-20.txt" 
annots = read.delim(f, na.strings = "", fill=TRUE, header=T, sep="\t") 
annots=annots[,c(1,5,6)] #Save selected (columns 1,5,6) info in annots table
```

Select top genes from each comparison:
#Select the most sigificantly different genes. 
```{r}
N=dim(eset)[1] 
top.M6PR.TvU = topTable(fit3,coef=1,number=N,p.value=0.05) 
top.COL.UvT = topTable(fit3,coef=2,number=N,p.value=0.05) 
top.Diff = topTable(fit3,coef=3,number=N,p.value=0.05) 
```

Merge annotations with top lists:
#So that we have annotation names instead of probe set numbers. 
```{r}
row.names(annots)=annots$array_element_name 
top.COL.UvT.merged=merge(annots,top.COL.UvT,by="row.names") #(A) Pathways mainly affected by salt stress
top.M6PR.TvU.merged=merge(annots,top.M6PR.TvU,by="row.names") #(B) Pathways mainly activated by the M6PR transgene
top.Diff.merged=merge(annots,top.Diff,by="row.names") #(C) Pathways affected by both salt stress and the M6PR transgene
```

Count differently expressed genes:
#because there may be multiple probe sets to one gene (annotation)
```{r}
genes.M6PR.TvU = unique(top.M6PR.TvU.merged$locus)
genes.COL.UvT = unique(top.COL.UvT.merged$locus)
genes.Diff = unique(top.Diff.merged$locus)
intersection = intersect(genes.COL.UvT,genes.M6PR.TvU)
```

CONCLUSION:











